src/Parser/Parser.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace UnicornFail\Emoji\Parser;
+
+use UnicornFail\Emoji\Environment\EmojiEnvironmentInterface;
+use UnicornFail\Emoji\Environment\Environment;
+use UnicornFail\Emoji\Event\DocumentParsedEvent;
+use UnicornFail\Emoji\Event\DocumentPreParsedEvent;
+use UnicornFail\Emoji\Input\Input;
+use UnicornFail\Emoji\Node\Block\Document;
+use UnicornFail\Emoji\Node\Inline\Emoticon;
+use UnicornFail\Emoji\Node\Inline\HtmlEntity;
+use UnicornFail\Emoji\Node\Inline\Shortcode;
+use UnicornFail\Emoji\Node\Inline\Text;
+use UnicornFail\Emoji\Node\Inline\Unicode;
+
+class Parser implements ParserInterface
```

Comment:
I haven't dug deeply into the parsing/lexing implementation yet so I don't have any comments here yet, other than that representing the parsed tokens as a flat "stream" (where each item in that "stream" represents either a single character or a group of characters) aligns very nicely with how the lexing works and how parsing UTF-8 words.  I therefore think we can keep the abstraction fairly straight-forward.

Reply:
This has been simplified and the `Input` class was removed.